nvidia/gpu.go

```diff
@@ -18,81 +18,129 @@
 package nvidia
 
 import (
-	"encoding/csv"
+	"encoding/xml"
 	"errors"
-	"io"
 	"os/exec"
+	"reflect"
 	"strconv"
 	"strings"
 
 	"github.com/elastic/beats/libbeat/common"
 )
 
-//GPUUtilization provides interface to utilization metrics and state of GPU.
-type GPUUtilization interface {
-	command(env string) *exec.Cmd
-	run(cmd *exec.Cmd, gpuCount int, query string, action Action) ([]common.MapStr, error)
-}
-
-//Utilization implements one flavour of GPUCount interface.
+// Utilization implements one flavour of GPUCount interface.
 type Utilization struct {
 }
 
-//newUtilization returns instance of Utilization
-func newUtilization() Utilization {
+// NewUtilization returns instance of Utilization
+func NewUtilization() Utilization {
 	return Utilization{}
 }
 
-func (g Utilization) command(env string, query string) *exec.Cmd {
-	if env == "test" {
-		return exec.Command("localnvidiasmi")
+func (g Utilization) command() *exec.Cmd {
+	return exec.Command("nvidia-smi", "-q", "-x")
+}
+
+// TrimmedInt allows for an Unmarshal which will crop off unit types
+type TrimmedInt int64
+
+// UnmarshalXML for the TrimmedInt type will trim off units from the string
+// e.g. "48 MB" becomes "48"
+func (t *TrimmedInt) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	var v string
+	d.DecodeElement(&v, &start)
+	strs := strings.Split(v, " ")
+
+	if len(strs) == 0 {
+		return errors.New("No values in string " + v)
+	}
+
+	i, err := strconv.ParseInt(strs[0], 10, 64)
+	if err != nil {
+		return err
+	}
+
+	*t = TrimmedInt(i)
+
+	return nil
+}
+
+// PopulateArgs takes in a reflected type and adds the field and its value
+// to the libbeat map if it has been configured to add it in the Map
+func PopulateArgs(m Map, val reflect.Value, event *common.MapStr) {
+	for i := 0; i < val.Type().NumField(); i++ {
+		valueField := val.Field(i)
+		typeField := val.Type().Field(i)
```

Comment:
It's probably worth deduplicating the `Type()` call to save a tiny bit of work https://golang.org/src/reflect/value.go?s=56202:56228#L1864